README.md
# Livesport TV Utils

[![Go Reference][goreference-img]][goreference]
[![Latest release][release-img]][release]
[![License: MIT][license-img]][license]
[![tests][tests-img]][tests]
[![Go Report Card][goreportcard-img]][goreportcard]
[![codecov][codecov-img]][codecov]

Open source implementation of Livesport TV utilities library.

Packages provide `Formatter` and `Parser` variables to override default behavior of
format methods (`MarshalText`, `MarshalJSON`, `String`, ...)
and parse methods (`UnmarshalText`, `UnmarshalJSON`, ...).
Default value of `Formatter` is `DefaultFormatter` and default value of `Parser` is `DefaultParser`.
Also, specific verbs for `fmt` formatting are available.

## Date
```go
import "go.lstv.dev/util/date"
```

- Type `Date` represents date (year, month, day).
- Function `New` to create new date.
- Function `DateFromTime` to create date from `time.Time`.
- Type `DateFilter` to work with date intervals and filtering.

## Roman
```go
import "go.lstv.dev/util/roman"
```

- Type `Number` represents roman number.
- It supports roman numerals:
  - `M` (1000)
  - `D` (500)
  - `C` (100)
  - `L` (50)
  - `X` (10)
  - `V` (5)
  - `I` (1)
- Short and long forms are also supported and configurable.

## Sem
```go
import "go.lstv.dev/util/sem"
```

- Type `Version` represents semantic version.
- Functions:
  - `Compare`, `CompareVersion` and `CompareTag`
  - `Latest`, `LatestVersion` and `LatestTag`
  - `Parse`, `ParseVersion` and `ParseTag`
- See [Semantic Versioning 2.0.0](https://semver.org/spec/v2.0.0.html) for more details.

## Size
```go
import "go.lstv.dev/util/size"
```

- Provides type `Size` to keep, marshal and unmarshal times-byte size values.
- It supports three JSON forms:
  - numeric form (JSON number is always in bytes)
  - string form (JSON string with or without units)
  - object form (JSON object like `{"value":1000,"unit":"MiB"}`)

## Test
```go
import "go.lstv.dev/util/test"
```

- Provides functions to test marshal and unmarshal methods:
  - `MarshalBinary` and `UnmarshalBinary`
  - `MarshalText` and `UnmarshalText`
  - `MarshalJSON` and `UnmarshalJSON`
- See [examples](./test/README.md).

## UUID
```go
import "go.lstv.dev/util/uu"
```

- Provides type `ID` to keep, marshal and unmarshal UUID values.

[codecov]: https://codecov.io/gh/livesport-tv/util
[codecov-img]: https://codecov.io/gh/livesport-tv/util/branch/master/graph/badge.svg?token=SPG1OPCHFF
[goreference]: https://pkg.go.dev/go.lstv.dev/util
[goreference-img]: https://pkg.go.dev/badge/go.lstv.dev/util.svg
[goreportcard]: https://goreportcard.com/report/go.lstv.dev/util
[goreportcard-img]: https://goreportcard.com/badge/go.lstv.dev/util
[license]: https://opensource.org/licenses/MIT
[license-img]: https://img.shields.io/github/license/livesport-tv/util
[release]: https://github.com/livesport-tv/util/releases
[release-img]: https://img.shields.io/github/v/release/livesport-tv/util?display_name=tag&sort=semver
[tests]: https://github.com/livesport-tv/util/actions/workflows/tests.yml
[tests-img]: https://github.com/livesport-tv/util/actions/workflows/tests.yml/badge.svg


CHANGELOG.md
# Changelog

## [Unreleased]

## [0.8.0] - 2022-05-14
### Added
- Panic handling for `test.*` functions.
- Field `Before` and `After` to `test.Case*` types.
- New error checkers:
  - `test.AnyError`
  - `test.ErrorHasPrefix`
  - `test.ErrorHasSuffix`
  - `test.ErrorMatch`

## [0.7.1] - 2022-04-27
### Fixed
- Package `test` allows to test also pointer types.

## [0.7.0] - 2022-04-25
### Added
- Format/parse and `fmt.Formatter` support for `uu` package.
- Function `uu.RandomID`.

### Fixed
- Error messages.

## [0.6.0] - 2022-04-19
### Added
- Package `uu` with UUID support.

## [0.5.0] - 2022-03-23
### Added
- Format `roman.FormatLowerCase`.
- Methods `sem.Ver.Core` and `sem.Ver.IsZero`.
- Implementation of `fmt.Formatter` for:
  - `date.Date`
  - `roman.Number`
  - `sem.Ver`
- Package `test` for enhanced testing.

### Changed
- Implementation of tests.

### Fixed
- Method `Date.MarshalBinary` error.

## [0.4.0] - 2022-03-16
### Added
- Package `constraint` with generic support and helpers.
- Function `size.Bytes` with generic support.
- Methods `sem.Ver.NextMajor`, `sem.Ver.NextMinor` and `sem.Ver.NextPatch`.

### Changed
- Type `Rule` doc comments (all packages).
- Used `any` instead of `interface{}`.
- Generic support:
  - Error types.
  - Function `size.New`
  - All parser functions now accepts `constraint.ParserInput`.
- Updated:
  - Go to 1.18
  - `github.com/stretchr/testify` to `1.7.1`

### Removed
- Methods of `size.Size`:
  - `BytesInt`
  - `BytesUint`
  - `BytesInt32`
  - `BytesUint32`
  - `BytesInt64`
  - `BytesUint64`
  - `BytesFloat32`
  - `BytesFloat64`

## [0.3.0] - 2022-02-28
### Added
- Type `Rule` and `DefaultParser` parameter of same type (all packages).

### Changed
- Renamed all unmarshal global functions to parse (all packages).
  - Format/Parse and Formatter/Parser is better choice than Format/UnmarshalText.
  - Variables `UnmarshalText` renamed to `Parser`.
  - Functions `DefaultUnmarshalText` renamed to `DefaultParser`.
- Published under-laying errors (all packages).
- Unified `input` and `data` names.
  - Renamed `MaxTextLength` to `MaxInputLength`.

### Removed
- Global unmarshal flag variables (all packages).

### Fixed
- Doc comments.

## [0.2.1] - 2022-02-25
### Fixed
- Added doc comments for `size.ObjectKeyValue` and `size.ObjectKeyUnit`.
- Method `sem.Ver.Compare` and function `sem.DefaultComparePreRelease` comparison of release and pre-release.

## [0.2.0] - 2022-02-24
### Added
- Introduced new constants `size.ObjectKeyValue` and `size.ObjectKeyUnit`.
- Missing doc comments.

### Changed
- Renamed `size.NewSize` to `size.New`.
- Optimized `size.unmarshalJSONObject` implementation.

### Removed
- Variable `sem.Zero`.
  - Use `sem.Ver{}` instead.
- Name of `date.filterNo.Contains` receiver.

## [0.1.0] - 2022-02-22
### Added
- First release of util.

[Unreleased]: https://github.com/livesport-tv/util/compare/v0.8.0...master
[0.8.0]: https://github.com/livesport-tv/util/compare/v0.7.1...v0.8.0
[0.7.1]: https://github.com/livesport-tv/util/compare/v0.7.0...v0.7.1
[0.7.0]: https://github.com/livesport-tv/util/compare/v0.6.0...v0.7.0
[0.6.0]: https://github.com/livesport-tv/util/compare/v0.5.0...v0.6.0
[0.5.0]: https://github.com/livesport-tv/util/compare/v0.4.0...v0.5.0
[0.4.0]: https://github.com/livesport-tv/util/compare/v0.3.0...v0.4.0
[0.3.0]: https://github.com/livesport-tv/util/compare/v0.2.1...v0.3.0
[0.2.1]: https://github.com/livesport-tv/util/compare/v0.2.0...v0.2.1
[0.2.0]: https://github.com/livesport-tv/util/compare/v0.1.0...v0.2.0
[0.1.0]: https://github.com/livesport-tv/util/releases/tag/v0.1.0


uu/parse.go
// Copyright 2022 Livesport TV s.r.o. All rights reserved.
// Use of this source code is governed by a MIT license
// that can be found in the LICENSE file.

package uu

import (
	"fmt"

	"go.lstv.dev/util/constraint"
)

var starts = []int{0, 2, 4, 6, 9, 11, 14, 16, 19, 21, 24, 26, 28, 30, 32, 34}

var (
	// MaxInputLength allows limiting DefaultParser input.
	// Set 0 to disable this setting.
	// ErrInputTooLong is wrapped and used if limit is exceeded.
	// Note: For year > 9999, increment MaxInputLength to > 45.
	MaxInputLength = 45

	// Parser is used by ID.UnmarshalText function.
	Parser = DefaultParser[[]byte]
)

type (
	// Rule allows configuring Parser behavior.
	// Available rules are:
	//   RuleDisableURN
	//   RuleDisableUpperCaseDigits
	Rule int
)

const (
	// RuleDisableURN disallow URN form.
	RuleDisableURN = Rule(1 << iota)

	// RuleDisableUpperCaseDigits disallow upper-case digits.
	RuleDisableUpperCaseDigits
)

// DefaultParser parse UUID from input.
//
// See also MaxInputLength.
func DefaultParser[T constraint.ParserInput](input T, r Rule) (id ID, err error) {
	const funcName = "DefaultParser"
	l := len(input)
	if MaxInputLength != 0 && l > MaxInputLength {
		// do not use input for "input too long" error
		var t T
		return ID{}, newParseError(funcName, t, fmt.Errorf("%w: %d > %d", ErrInputTooLong, l, MaxInputLength))
	}
	offset := 0
	switch l {
	case IDLength:
		// no-op
	case IDLength + len(URNPrefix):
		if r&RuleDisableURN != 0 {
			return ID{}, newParseError(funcName, input, ErrURNFormatDisabled)
		}
		if !hasURNPrefix(input) {
			return ID{}, newParseError(funcName, input, nil)
		}
		// move offset after URN prefix
		offset = len(URNPrefix)
	default:
		return ID{}, newParseError(funcName, input, nil)
	}
	if input[offset+8] != '-' || input[offset+13] != '-' || input[offset+18] != '-' || input[offset+23] != '-' {
		return ID{}, newParseError(funcName, input, nil)
	}
	n := [2]uint64{}
	allowUpperCase := r&RuleDisableUpperCaseDigits == 0
	for i, start := range starts {
		for j := 0; j < 2; j++ {
			if v, ok := parseDigit(input[offset+start+j], allowUpperCase); ok {
				x := 124 - (i*2+j)*4
				n[x>>6] |= v << (x & 0x3f)
				continue
			}
			return ID{}, newParseError(funcName, input, InvalidDigitError(input[offset+start+j]))
		}
	}
	return ID{
		Higher: n[1],
		Lower:  n[0],
	}, nil
}

func hasURNPrefix[T constraint.ParserInput](input T) bool {
	if input[0] != 'u' && input[0] != 'U' {
		return false
	}
	if input[1] != 'r' && input[1] != 'R' {
		return false
	}
	if input[2] != 'n' && input[2] != 'N' {
		return false
	}
	for i, r := range []byte(":uuid:") {
		if input[i+3] != r {
			return false
		}
	}
	return true
}

func parseDigit(digit byte, allowUpperCase bool) (uint64, bool) {
	if digit >= '0' && digit <= '9' {
		return uint64(digit - '0'), true
	}
	if digit >= 'a' && digit <= 'f' {
		return uint64(digit - ('a' - 10)), true
	}
	if allowUpperCase && digit >= 'A' && digit <= 'F' {
		return uint64(digit - ('A' - 10)), true
	}
	return 0, false
}
